process/block/shardblockProposal.go

```diff
@@ -0,0 +1,257 @@
+func (sp *shardProcessor) CreateBlockProposal(
+	initialHdr data.HeaderHandler,
+	haveTime func() bool,
+) (data.HeaderHandler, data.BodyHandler, error) {
+	if check.IfNil(initialHdr) {
+		return nil, nil, process.ErrNilBlockHeader
+	}
+	shardHdr, ok := initialHdr.(data.ShardHeaderHandler)
+	if !ok {
+		return nil, nil, process.ErrWrongTypeAssertion
+	}
+
+	err := sp.updateEpochIfNeeded(shardHdr)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	sp.miniBlocksSelectionSession.ResetSelectionSession()
+	err = sp.createBlockBodyProposal(shardHdr, haveTime)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// todo: check empty mini blocks vs nil. Same for block.Body.MiniBlocks
+	err = shardHdr.SetMiniBlockHeaderHandlers(sp.miniBlocksSelectionSession.GetMiniBlockHeaderHandlers())
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// todo: check also the rest of the header fields (e.g what was previously done on applyBodyToHeader)
+
+	return shardHdr, &block.Body{MiniBlocks: sp.miniBlocksSelectionSession.GetMiniBlocks()}, nil
+}
+
+func (sp *shardProcessor) createBlockBodyProposal(
+	shardHdr data.HeaderHandler,
+	haveTime func() bool,
+) error {
+	sp.blockSizeThrottler.ComputeCurrentMaxSize()
+
+	log.Debug("started creating block body",
+		"epoch", shardHdr.GetEpoch(),
+		"round", shardHdr.GetRound(),
+		"nonce", shardHdr.GetNonce(),
+	)
+
+	return sp.createProposalMiniBlocks(haveTime)
+}
+
+func (sp *shardProcessor) selectIncomingMiniBlocksForProposal(
+	haveTime func() bool,
+) error {
+	log.Debug("selectIncomingMiniBlocksForProposal has been started")
+
+	sw := core.NewStopWatch()
+	sw.Start("ComputeLongestMetaChainFromLastNotarized")
+	orderedMetaBlocks, orderedMetaBlocksHashes, err := sp.blockTracker.ComputeLongestMetaChainFromLastNotarized()
+	sw.Stop("ComputeLongestMetaChainFromLastNotarized")
+	log.Debug("measurements", sw.GetMeasurements()...)
+	if err != nil {
+		return err
+	}
+
+	log.Debug("meta blocks ordered", "num meta blocks", len(orderedMetaBlocks))
+
+	lastMetaHdr, _, err := sp.blockTracker.GetLastCrossNotarizedHeader(core.MetachainShardId)
+	if err != nil {
+		return err
+	}
+
+	err = sp.selectIncomingMiniBlocks(lastMetaHdr, orderedMetaBlocks, orderedMetaBlocksHashes, haveTime)
+	if err != nil {
+		return err
+	}
+
+	referencedMetaBlocks := sp.miniBlocksSelectionSession.GetReferencedMetaBlocks()
+	numHeadersAdded := uint32(len(referencedMetaBlocks))
+	if numHeadersAdded > 0 {
+		go sp.requestMetaHeadersIfNeeded(numHeadersAdded, referencedMetaBlocks[numHeadersAdded-1])
+	}
+
+	miniBlockHeaderHandlers := sp.miniBlocksSelectionSession.GetMiniBlockHeaderHandlers()
+	for _, miniBlockHeader := range miniBlockHeaderHandlers {
+		log.Debug("mini block info",
+			"type", miniBlockHeader.GetTypeInt32(),
+			"sender shard", miniBlockHeader.GetSenderShardID(),
+			"receiver shard", miniBlockHeader.GetReceiverShardID(),
+			"txs added", miniBlockHeader.GetTxCount())
+	}
```

Comment:
is this for loop needed? isn't it more or less duplicate of [this one](https://github.com/multiversx/mx-chain-go/blob/create-block-proposal/process/coordinator/processProposal.go#L104) that happens earlier?